### ISSUE TITLE
[FEA]: (Work in Progress) Add c++ level 'DataManager', and high level python 'PayloadManager' objects.

### ISSUE THREAD

**Issue by drobison00** [nv-morpheus/Morpheus#1206]
### Is this a new feature, an improvement, or a change to existing functionality?

New Feature

### How would you describe the priority of this feature request

High

### Please provide a clear description of problem this feature solves

This feature will support three primary goals:

- Seamlessly support existing ControlMessage use cases, with no performance degradation.
- Introduce new c++ DataRecordManager object which is responsible for CRUD operations on DataRecord objects providing the same basic CRUD interface for a specific type of backing device.
- Introduce new PayloadManager object which will take the place of the existing ControlMessage payload; this class will be responsible for keeping track of stored object metadata.

### Describe your ideal solution

# Pseudocode
## Data Manager Object

`Management Object`
```c++
class DataRecordManager {
public:
  // Constructors & Destructors
  DataRecordManager();
  ~DataRecordManager();

  // CRUD Operations
  bool CreateRecord(const std::string& key, const std::vector<uint8_t>& data, const std::string& type);
  bool ReadRecord(const std::string& key, std::vector<uint8_t>& data_out);
  bool UpdateRecord(const std::string& key, const std::vector<uint8_t>& new_data);
  bool DeleteRecord(const std::string& key);

private:
  std::unordered_map<std::string, std::unique_ptr<DataRecord>> data_map_;
};
```

`Base DataRecord class`
```c++
class DataRecord {
public:
  virtual ~DataRecord() = default;
  virtual bool Create(const std::vector<uint8_t>& data) = 0;
  virtual bool Read(std::vector<uint8_t>& data_out) = 0;
  virtual bool Update(const std::vector<uint8_t>& new_data) = 0;
  virtual bool Delete() = 0;
};
```

`Derived in-memory class`
```c++
class MemoryDataRecord : public DataRecord {
public:
  bool Create(const std::vector<uint8_t>& data) override;
  bool Read(std::vector<uint8_t>& data_out) override;
  bool Update(const std::vector<uint8_t>& new_data) override;
  bool Delete() override;

private:
  std::vector<uint8_t> data_;
};
```

`Derived on-disk record class`
```c++
class DiskDataRecord : public DataRecord {
public:
  bool Create(const std::vector<uint8_t>& data) override;
  bool Read(std::vector<uint8_t>& data_out) override;
  bool Update(const std::vector<uint8_t>& new_data) override;
  bool Delete() override;
private:
 std::string m_file_path;
};
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct
- [X] I have searched the [open feature requests](https://github.com/nv-morpheus/Morpheus/issues?q=is%3Aopen+is%3Aissue+label%3A%22feature+request%22%2Cimprovement%2Cenhancement) and have found no duplicates for this feature request

**Comment by cwharris** [comment]
I _think_ i understand how it would translate, but I'd really like to see an example that's framed around a stage, specifically the map operation of the stage (where by "map" I mean `self.process_message` in `mrc.operators.map(self.process_message))`).

**Comment by drobison00** [comment]
@cwharris I think I might need a bit more context on your question. In the context of usage in a stage, here is a snippet from the dfp_training code, and how it would be adapted to produce the same behavior as before:

`Before`
```python
        while (control_message.has_task("training")):
            control_message.remove_task("training")

            user_id = control_message.get_metadata("user_id")
            message_meta = control_message.payload()

            with message_meta.mutable_dataframe() as dfm:
                final_df = dfm.to_pandas()

            model = AutoEncoder(**model_kwargs)

            # Only train on the feature columns
            train_df = final_df[final_df.columns.intersection(feature_columns)]
            model.fit(train_df)
```
`After`
```python
        while (control_message.has_task("training")):
            control_message.remove_task("training")

            user_id = control_message.get_metadata("user_id")
            final_df = control_message.payload().read()

            model = AutoEncoder(**model_kwargs)

            # Only train on the feature columns
            train_df = final_df[final_df.columns.intersection(feature_column s)]
            model.fit(train_df)
```

In the context of having a very large, backed data source, something like this would probably be more appropriate:
`After`
```python
        while (control_message.has_task("training")):
            control_message.remove_task("training")

            user_id = control_message.get_metadata("user_id")
            df_loader = control_message.payload().as_data_loader(feature_columns) # Note, this isn't in the pseudo code.

            model = AutoEncoder(**model_kwargs)

            # Only train on the feature columns
            model.fit(df_loader)
```

**Comment by cwharris** [comment]
Question: Since reading/writing might require IO, would it be beneficial to have async versions of the CRUD operations? On the C++ side, I imagine we'd release the gil. That allows other threads to continue processing, but without an async API in Python, we're blocking the current thread from reading more messages while any CRUD is in-flight.

**Comment by drobison00** [comment]
@cwharris Yes, for most operations we'll likely want to drop the GIL or perform them asynchronously. I won't say this is 100% true, because for some record managers, we may utilize something like fsspec via pybind for the initial implementation.

In the context of a python call, we should see other python threads have a chance to grab the GIL and start executing as soon as we drop it in the c++ function. Similarly, when the c++ function issues a syscall for something like a file or network read, we should see it yield as well and let some other thread execute.

If I'm understanding the context correctly, the only thing we're likely to be blocked on is processing more messages within the current python node; I don't know of any current situations where this would be a problem.
